bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go

```diff
@@ -94,6 +97,33 @@ type KubeadmConfigSpec struct {
 	// For more information, refer to https://github.com/kubernetes-sigs/cluster-api/pull/2763#discussion_r397306055.
 	// +optional
 	UseExperimentalRetryJoin bool `json:"useExperimentalRetryJoin,omitempty"`
+
+	// Ignition contains Ignition specific configuration.
```

Comment:
I think my only remaining question is about the API changes:
am I right assuming that this is not needed for the cluster to bootstrap successfully (only need internal templates) but only to add additional OS customisations?

If so couldn't we find and alternative which does not impact the API until we remove the feature gate or preferably we proceed with https://github.com/kubernetes-sigs/cluster-api/issues/5294?

e.g Ignition Format could be an annotation and AdditionalConfig could live in a well known user pre-populated cm/secret that the controller fetches and the template incorporate via ign append stanza.

Thoughts?

Reply:
I guess we could put both the user input and possibly Ignition output in annotations if that's preferable, it shouldn't be much of a change either. Would annotation data be also behind the feature gate then?

Are there some other examples of such thing in CAPI ecosystem? I'm aware of the examples in K8s in general.

Reply:
Kubernetes itself deprecated using annotations for alpha fields:

> In early versions of Kubernetes, some in-development features represented new API fields as annotations, generally with the form something.alpha.kubernetes.io/name or something.beta.kubernetes.io/name (depending on our confidence in it). This pattern is deprecated. Some such annotations may still exist, but no new annotations may be defined. New API fields are now developed as regular fields.
> https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#label-selector-and-annotation-conventions

Reply:
Annotations are not really even needed. AdditionalConfig could just live in a well known user pre-populated cm/secret. If present the controller fetches it and the template incorporate via ign append stanza and go ign path. Just a thought so we keep the API untouched for now. Happy to hear from others and go with the approach that fits for everyone :)

Reply:
> I think my only remaining question is about the API changes: am I right assuming that this is not needed for the cluster to bootstrap successfully (just need internal templates) but only to add additional OS customisations?

I'd say this isn't the case, no. In the case of Flatcar for example, there are some distro-specific configuration without which the cluster won't bootstrap and we pass this configuration in `AdditionalConfig`. I don't think we want to include this configuration in the Ignition implementation because doing so would make it work only for Flatcar (right @invidian?).

Given the above, does it still make sense to move the contents of `AdditionalConfig` to a CM or a secret?

Reply:
@johananl is right. `AdditionalConfig` will contain both platform-specific and Flatcar-specific bits of configuration, while configuration in the bootstrap provider aims to be as agnostic as possible. Right now both CAPA and CAPZ templates use `AdditionalConfig` config field.

Reply:
ok I see, I think that's unfortunate but not a blocker from my side for the feature gated solution. We'll figure a more polish workflow with https://github.com/kubernetes-sigs/cluster-api/issues/5294